# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/ContentParent.cpp
# Commit: 6fd57b83bb52
# Full Hash: 6fd57b83bb52e9683bade8cbfb5f8ba6575ac6af
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-02-07 03:55:51
# Regressor Bug: 1582832
# File Overlap Count: 3
# Description:
#   Bug 1582832: Part 1 - Make FrameLoader owner rather than DocShell responsible for discarding a BC. r=nika
#   
#   There are all sorts of lifecycle issues which arise from making DocShell
#   responsible for discarding BrowsingContexts. In this particular bug, we tend
#   to run into them in cases where we create a BrowsingContext for a FrameLoader,
# ==============================================================================

diff -r 28a9e4b075af -r 6fd57b83bb52 dom/ipc/ContentParent.cpp
--- a/dom/ipc/ContentParent.cpp	Thu Feb 06 18:35:10 2020 +0000
+++ b/dom/ipc/ContentParent.cpp	Thu Feb 06 19:07:56 2020 +0000
@@ -6150,6 +6150,17 @@
   return true;
 }
 
+bool ContentParent::CheckBrowsingContextEmbedder(BrowsingContext* aBC,
+                                                 const char* aOperation) const {
+  if (!aBC->Canonical()->IsEmbeddedInProcess(ChildID())) {
+    MOZ_LOG(BrowsingContext::GetLog(), LogLevel::Warning,
+            ("ParentIPC: Trying to %s out of process context 0x%08" PRIx64,
+             aOperation, aBC->Id()));
+    return false;
+  }
+  return true;
+}
+
 mozilla::ipc::IPCResult ContentParent::RecvDetachBrowsingContext(
     uint64_t aContextId, DetachBrowsingContextResolver&& aResolve) {
   // NOTE: Immediately resolve the promise, as we've received the message. This
@@ -6165,24 +6176,12 @@
     return IPC_OK();
   }
 
-  if (!CheckBrowsingContextOwnership(context, "detach")) {
-    // We're trying to detach a child BrowsingContext in another child
-    // process. This is illegal since the owner of the BrowsingContext
-    // is the proccess with the in-process docshell, which is tracked
-    // by OwnerProcessId.
-    return IPC_OK();
+  if (!CheckBrowsingContextEmbedder(context, "detach")) {
+    return IPC_FAIL(this, "Illegal Detach() attempt");
   }
 
   context->Detach(/* aFromIPC */ true);
 
-  context->Group()->EachOtherParent(this, [&](ContentParent* aParent) {
-    // Hold a reference to `context` until the response comes back to ensure it
-    // doesn't die while messages relating to this context are in-flight.
-    auto resolve = [context](bool) {};
-    auto reject = [context](ResponseRejectReason) {};
-    aParent->SendDetachBrowsingContext(context->Id(), resolve, reject);
-  });
-
   return IPC_OK();
 }
 
@@ -6320,6 +6319,13 @@
     return IPC_OK();
   }
 
+  if (aData.source() && aData.source()->IsDiscarded()) {
+    MOZ_LOG(
+        BrowsingContext::GetLog(), LogLevel::Debug,
+        ("ParentIPC: Trying to send a message from dead or detached context"));
+    return IPC_OK();
+  }
+
   RefPtr<ContentParent> cp = aContext->Canonical()->GetContentParent();
   if (!cp) {
     MOZ_LOG(BrowsingContext::GetLog(), LogLevel::Debug,